File: backend/api/schemas/drafts.py
"""
Draft API schemas — Phase 2N.9.

Request and response models for the /drafts REST endpoints.
These are passive data-transfer types only. No execution or validation logic.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from backend.tools.toolset import StrategyToolSet


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class DraftCreateRequest(BaseModel):
    """Request body for POST /drafts."""

    model_config = ConfigDict(extra="forbid")

    draft_id: str
    display_name: str
    description: str | None = None
    toolset: StrategyToolSet
    enabled: bool = True
    tags: list[str] = []
    notes: str | None = None


class DraftUpdateRequest(BaseModel):
    """
    Request body for PUT /drafts/{draft_id}.

    All fields are optional. Only explicitly provided fields are applied.
    Omitting a field leaves the existing value unchanged.
    """

    model_config = ConfigDict(extra="forbid")

    display_name: str | None = None
    description: str | None = None
    toolset: StrategyToolSet | None = None
    enabled: bool | None = None
    tags: list[str] | None = None
    notes: str | None = None


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class DraftToolConfigResponse(BaseModel):
    instance_id: str
    tool_id: str
    parameters: dict[str, Any]
    enabled: bool
    display_name: str | None
    color: str | None


class DraftToolSetResponse(BaseModel):
    toolset_id: str
    display_name: str | None
    tools: list[DraftToolConfigResponse]
    enabled: bool


class DraftResponse(BaseModel):
    draft_id: str
    display_name: str
    description: str | None
    toolset: DraftToolSetResponse
    created_at: datetime
    updated_at: datetime
    enabled: bool
    tags: list[str]
    notes: str | None


class DraftListResponse(BaseModel):
    drafts: list[DraftResponse]
    count: int

File: frontend/src/api/drafts.ts
/**
 * Typed API client for all /drafts backend endpoints.
 *
 * No business logic. No validation duplication.
 * Backend is always the source-of-truth — this layer is pure HTTP transport.
 */
import type {
  CompositionValidationResponse,
  DraftListResponse,
  StrategyDraftData,
} from '../types/drafts'

export type { CompositionValidationResponse, DraftListResponse, StrategyDraftData }

// ---------------------------------------------------------------------------
// Request shapes (mirror backend schemas)
// ---------------------------------------------------------------------------

export interface DraftCreateRequest {
  draft_id: string
  display_name: string
  description?: string | null
  toolset: {
    toolset_id: string
    tools: ToolConfigPayload[]
    display_name?: string | null
    enabled?: boolean
  }
  enabled?: boolean
  tags?: string[]
  notes?: string | null
}

export interface DraftUpdateRequest {
  display_name?: string
  description?: string | null
  enabled?: boolean
  tags?: string[]
  notes?: string | null
}

export interface ToolConfigPayload {
  instance_id: string
  tool_id: string
  parameters: Record<string, unknown>
  enabled?: boolean
  display_name?: string | null
  color?: string | null
}

export interface AddToolRequest {
  tool: ToolConfigPayload
  index?: number | null
}

export interface PatchToolRequest {
  parameters?: Record<string, unknown>
  enabled?: boolean
  display_name?: string | null
  color?: string | null
}

export interface ReorderToolsRequest {
  ordered_instance_ids: string[]
}

// ---------------------------------------------------------------------------
// HTTP helper
// ---------------------------------------------------------------------------

async function _req<T>(
  method: string,
  path: string,
  body?: unknown,
): Promise<T> {
  const resp = await fetch(path, {
    method,
    headers: body !== undefined ? { 'Content-Type': 'application/json' } : {},
    body: body !== undefined ? JSON.stringify(body) : undefined,
  })

  // 204 No Content — void operations (delete, archive)
  if (resp.status === 204) return undefined as unknown as T

  const data = await resp.json().catch(() => ({ detail: resp.statusText }))

  if (!resp.ok) {
    const detail = (data as { detail?: unknown }).detail
    const msg =
      typeof detail === 'string'
        ? detail
        : JSON.stringify(detail ?? `HTTP ${resp.status}`)
    throw new Error(msg)
  }

  return data as T
}

// ---------------------------------------------------------------------------
// Draft CRUD
// ---------------------------------------------------------------------------

export function fetchDrafts(): Promise<DraftListResponse> {
  return _req('GET', '/drafts')
}

export function fetchDraft(id: string): Promise<StrategyDraftData> {
  return _req('GET', `/drafts/${id}`)
}

export function createDraft(
  payload: DraftCreateRequest,
): Promise<StrategyDraftData> {
  return _req('POST', '/drafts', payload)
}

export function updateDraft(
  id: string,
  patch: DraftUpdateRequest,
): Promise<StrategyDraftData> {
  return _req('PUT', `/drafts/${id}`, patch)
}

export function deleteDraft(id: string): Promise<void> {
  return _req('DELETE', `/drafts/${id}`)
}

export function archiveDraft(id: string): Promise<void> {
  return _req('POST', `/drafts/${id}/archive`)
}

// ---------------------------------------------------------------------------
// Draft composition
// ---------------------------------------------------------------------------

export function addToolToDraft(
  id: string,
  req: AddToolRequest,
): Promise<StrategyDraftData> {
  return _req('POST', `/drafts/${id}/tools`, req)
}

export function removeToolFromDraft(
  id: string,
  instanceId: string,
): Promise<StrategyDraftData> {
  return _req('DELETE', `/drafts/${id}/tools/${instanceId}`)
}

export function reorderDraftTools(
  id: string,
  req: ReorderToolsRequest,
): Promise<StrategyDraftData> {
  return _req('POST', `/drafts/${id}/tools/reorder`, req)
}

export function patchDraftTool(
  id: string,
  instanceId: string,
  req: PatchToolRequest,
): Promise<StrategyDraftData> {
  return _req('PATCH', `/drafts/${id}/tools/${instanceId}`, req)
}

export function validateDraft(
  id: string,
): Promise<CompositionValidationResponse> {
  return _req('POST', `/drafts/${id}/validate`)
}
